rbx_types/src/attributes/writer.rs

```diff
@@ -0,0 +1,136 @@
+        .write_all(&n.to_le_bytes().to_vec())
+        .expect("couldn't write f32 to buffer");
+}
+
+fn write_u32<W: Write>(mut writer: W, n: u32) {
+    writer
+        .write_all(&n.to_le_bytes().to_vec())
+        .expect("couldn't write u32 to buffer");
+}
+
+fn write_color3<W: Write>(mut writer: W, color: Color3) {
+    write_f32(&mut writer, color.r);
+    write_f32(&mut writer, color.g);
+    write_f32(&mut writer, color.b);
+}
+
+fn write_string<T: AsRef<[u8]>, W: Write>(mut writer: W, string: T) {
+    let bytes = string.as_ref();
+    write_u32(&mut writer, bytes.len() as u32);
+    writer
+        .write_all(bytes)
+        .expect("couldn't write string to buffer");
+}
+
+fn write_udim<W: Write>(mut writer: W, udim: UDim) {
+    write_f32(&mut writer, udim.scale);
+    writer
+        .write_all(&udim.offset.to_le_bytes().to_vec())
+        .expect("couldn't write UDim to buffer");
+}
+
+fn write_vector2<W: Write>(mut writer: W, vector2: Vector2) {
+    write_f32(&mut writer, vector2.x);
+    write_f32(&mut writer, vector2.y);
+}
+
+/// Writes the attribute property (AttributesSerialize) from a map of attribute names -> values.
+pub(crate) fn attributes_from_map<
+    K: Borrow<String>,
+    V: Borrow<Variant>,
+    I: Iterator<Item = (K, V)> + ExactSizeIterator,
+    M: IntoIterator<IntoIter = I, Item = (K, V)>,
+>(
+    map: M,
+) -> Result<Vec<u8>, AttributeError> {
+    let map = map.into_iter();
+    let mut bytes = Vec::new();
+
+    bytes
+        .write_all(&(map.len() as u32).to_le_bytes())
+        .expect("couldn't write map length to buffer");
+
+    for (name, variant) in map {
+        let variant = variant.borrow();
+
+        write_string(&mut bytes, name.borrow());
+
+        let attribute_type = AttributeType::try_from(variant.ty())?;
+        bytes.push(attribute_type as u8);
+
+        match (attribute_type, variant) {
+            (AttributeType::Bool, Variant::Bool(bool)) => bytes.push(*bool as u8),
+            (AttributeType::BrickColor, Variant::BrickColor(color)) => {
+                write_u32(&mut bytes, *color as u32)
+            }
+            (AttributeType::Color3, Variant::Color3(color)) => write_color3(&mut bytes, *color),
+            (AttributeType::ColorSequence, Variant::ColorSequence(sequence)) => {
+                write_u32(&mut bytes, sequence.keypoints.len() as u32);
+
+                for keypoint in &sequence.keypoints {
+                    write_f32(&mut bytes, 0.0); // Envelope
+                    write_f32(&mut bytes, keypoint.time);
+                    write_color3(&mut bytes, keypoint.color);
+                }
+            }
+            (AttributeType::Float32, Variant::Float32(float)) => {
+                write_f32(&mut bytes, *float);
+            }
+            (AttributeType::Float64, Variant::Float64(float)) => {
+                bytes
+                    .write_all(&float.to_le_bytes().to_vec())
+                    .expect("couldn't write double to buffer");
```

Comment:
Please factor out this into its own method and adopt the advice from the other versions of it (ie avoiding allocating an 8 byte `Vec`).